[BUG] TypeError after replacing tools on a toolbar

My plot should have different sets of tools depending on what is displayed.
setting `plot.toolbar` doesn't help since the toolbar model change not not handled correctly.
Executing something like `plot.toolbar.tools = new_tools_list` seems to work but results in this exception in JS console:
```
bokeh.js:30582 Uncaught TypeError: Cannot read property 'el' of undefined
    at el (bokeh.js:30582)
    at Array.map (<anonymous>)
    at ToolbarBaseView.render (bokeh.js:30586)
    at ToolbarPanelView.render (bokeh.js:11764)
    at PlotView._paint_levels (bokeh.js:22064)
    at PlotView.paint (bokeh.js:22044)
    at PlotView.repaint (bokeh.js:21983)
    at PlotView.<anonymous> (bokeh.js:21892)
    at Signal0.Signal.emit (bokeh.js:4167)
    at Signal0.emit (bokeh.js:4180)
```

Both `Toolbar` and `ToolbarBaseView` attach a signal on the `tools` property change. However, the `Toolbar.gestures` property change does not remove the old references:
```
// Toolbar._init_tools()
if (!some(this.gestures[et].tools, (t) => t.id == tool.id))
  this.gestures[et].tools = this.gestures[et].tools.concat([tool])
```
Because of this, only the addition of tools works without errors. And while `gestures` is a property in BokehJS, it is not in the Python counterpart.

Similar code can be seen in `ProxyToolbar` but it doesn't handle the changes to the list of tools at all, so I guess it was never on the plate.

While it's not reflected in the error above, other attributes have the same issue of not being pruned on tools change: `inspectors`, `help`, `actions`.

OK. First of all, I hate TS. With passion. 95% of the time has been spent chasing obscure type errors. Understanding and debugging C++ w/boost templates compilation errors was more fun.

Second of all, the tests hate me.
For example, among dozens of other errors, I get this one:
```
11) ProxyToolbar
       _merge_tools method
         should avoid merge of multiple multi-gesture tools:

      AssertionError: expected 9 to equal 3
      + expected - actual

      -9
      +3
```
I have added
```
console.log("PROXY", proxy_toolbar.actions.length);
```
to the test, just above the
```
expect(proxy_toolbar.actions.length).equal(3)
```
And it outputs `PROXY 3`! WTH? Where on Earth does it get 9? I tried debugging - there's 3 just as well.

--- a/bokehjs/src/lib/models/tools/toolbar.ts
+++ b/bokehjs/src/lib/models/tools/toolbar.ts
@@ -1,11 +1,8 @@
 import * as p from "core/properties"
-import {logger} from "core/logging"
-import {isString, isArray} from "core/util/types"
-import {some, sort_by, includes} from "core/util/array"
+import {isArray} from "core/util/types"
+import {sort_by, includes, intersection} from "core/util/array"
 
 import {Tool} from "./tool"
-import {ActionTool} from "./actions/action_tool"
-import {HelpTool} from "./actions/help_tool"
 import {GestureTool} from "./gestures/gesture_tool"
 import {InspectTool} from "./inspectors/inspect_tool"
 
@@ -17,19 +14,39 @@ export type Inspection = Tool
 export type Scroll = Tool
 export type Tap = Tool
 
+type ActiveGestureToolsProps = {
+  active_drag: p.Property<Drag | "auto" | null>
+  active_scroll: p.Property<Scroll | "auto" | null>
+  active_tap: p.Property<Tap | "auto" | null>
+  active_multi: p.Property<GestureTool | null>
+}
+
+type ActiveToolsProps = ActiveGestureToolsProps & {
+  active_inspect: p.Property<Inspection | Inspection[] | "auto" | null>
+}
+
 export namespace Toolbar {
   export type Attrs = p.AttrsOf<Props>
 
-  export type Props = ToolbarBase.Props & {
-    active_drag: p.Property<Drag | "auto">
-    active_inspect: p.Property<Inspection | Inspection[] | "auto" | null>
-    active_scroll: p.Property<Scroll | "auto">
-    active_tap: p.Property<Tap | "auto">
-    active_multi: p.Property<GestureTool | null>
-  }
+  export type Props = ToolbarBase.Props & ActiveToolsProps
 }
 
 export interface Toolbar extends Toolbar.Attrs {}
+
+const _get_active_attr = (et: string): keyof ActiveGestureToolsProps | null => {
+  switch (et) {
+    case 'tap': return 'active_tap'
+    case 'pan': return 'active_drag'
+    case 'pinch':
+    case 'scroll': return 'active_scroll'
+    case 'multi': return 'active_multi'
+  }
+  return null
+}
+
+const _supports_auto = (et: string) => {
+  return et == 'tap' || et == 'pan'
+}
 
 export class Toolbar extends ToolbarBase {
   properties: Toolbar.Props
@@ -50,66 +67,32 @@ export class Toolbar extends ToolbarBase {
     })
   }
 
-  initialize(): void {
-    super.initialize()
-    this._init_tools()
-  }
-
   connect_signals(): void {
     super.connect_signals()
     this.connect(this.properties.tools.change, () => this._init_tools())
   }
 
   protected _init_tools(): void {
-    for (const tool of this.tools) {
-      if (tool instanceof InspectTool) {
-        if (!some(this.inspectors, (t) => t.id == tool.id)) {
-          this.inspectors = this.inspectors.concat([tool])
-        }
-      } else if (tool instanceof HelpTool) {
-        if (!some(this.help, (t) => t.id == tool.id)) {
-          this.help = this.help.concat([tool])
-        }
-      } else if (tool instanceof ActionTool) {
-        if (!some(this.actions, (t) => t.id == tool.id)) {
-          this.actions = this.actions.concat([tool])
-        }
-      } else if (tool instanceof GestureTool) {
-        let event_types: GestureType[]
-        let multi: boolean
-        if (isString(tool.event_type)) {
-          event_types = [tool.event_type]
-          multi = false
-        } else {
-          event_types = tool.event_type || []
-          multi = true
-        }
-
-        for (let et of event_types) {
-          if (!(et in this.gestures)) {
-            logger.warn(`Toolbar: unknown event type '${et}' for tool: ${tool.type} (${tool.id})`)
-            continue
-          }
-
-          if (multi)
-            et = "multi"
-
-          if (!some(this.gestures[et].tools, (t) => t.id == tool.id))
-            this.gestures[et].tools = this.gestures[et].tools.concat([tool])
-
-          this.connect(tool.properties.active.change, this._active_change.bind(this, tool))
-        }
-      }
-    }
+    super._init_tools()
 
     if (this.active_inspect == 'auto') {
       // do nothing as all tools are active be default
     } else if (this.active_inspect instanceof InspectTool) {
+      let found = false
       for (const inspector of this.inspectors) {
         if (inspector != this.active_inspect)
           inspector.active = false
+        else
+          found = true
+      }
+      if (!found) {
+        this.active_inspect = null
       }
     } else if (isArray(this.active_inspect)) {
+      const active_inspect = intersection(this.active_inspect, this.inspectors)
+      if (active_inspect.length != this.active_inspect.length) {
+        this.active_inspect = active_inspect
+      }
       for (const inspector of this.inspectors) {
         if (!includes(this.active_inspect, inspector))
           inspector.active = false
@@ -127,42 +110,33 @@ export class Toolbar extends ToolbarBase {
         tool.active = true
     }
 
+    // Connecting signals has to be done before changing the active state of the tools.
     for (const et in this.gestures) {
       const gesture = this.gestures[et as GestureType]
 
-      if (gesture.tools.length == 0)
-        continue
-
       gesture.tools = sort_by(gesture.tools, (tool) => tool.default_order)
-
-      if (et == 'tap') {
-        if (this.active_tap == null)
-          continue
-
-        if (this.active_tap == 'auto')
-          _activate_gesture(gesture.tools[0])
-        else
-          _activate_gesture(this.active_tap)
+      for (const tool of gesture.tools) {
+        this.connect(tool.properties.active.change, this._active_change.bind(this, tool))
       }
+    }
 
-      if (et == 'pan') {
-        if (this.active_drag == null)
-          continue
-
-        if (this.active_drag == 'auto')
-          _activate_gesture(gesture.tools[0])
-        else
-          _activate_gesture(this.active_drag)
+    for (const et in this.gestures) {
+      const active_attr = _get_active_attr(et)
+      if (active_attr) {
+        const active_tool = this[active_attr]
+        if (active_tool == 'auto') {
+          const gesture = this.gestures[et as GestureType]
+          if (gesture.tools.length != 0 && _supports_auto(et)) {
+            _activate_gesture(gesture.tools[0])
+          }
+        } else if (active_tool != null) {
+          if (includes(this.tools, active_tool)) {
+            _activate_gesture(active_tool)
+          } else {
+            this[active_attr] = null
+          }
+        }
       }
-
-      if (et == 'pinch' || et == 'scroll') {
-        if (this.active_scroll == null || this.active_scroll == 'auto')
-          continue
-        _activate_gesture(this.active_scroll)
-      }
-
-      if (this.active_multi != null)
-        _activate_gesture(this.active_multi)
     }
   }
 }

--- a/bokehjs/src/lib/models/tools/toolbar_base.ts
+++ b/bokehjs/src/lib/models/tools/toolbar_base.ts
@@ -5,6 +5,8 @@ import * as p from "core/properties"
 import {DOMView} from "core/dom_view"
 import {Logo, Location} from "core/enums"
 import {EventType} from "core/ui_events"
+import {some, every} from "core/util/array"
+import {Set} from "core/util/data_structures"
 import {isString} from "core/util/types"
 import {Model} from "model"
 import {Tool} from "./tool"
@@ -64,7 +66,10 @@ export class ToolbarBaseView extends DOMView {
 
   connect_signals(): void {
     super.connect_signals()
-    this.connect(this.model.properties.tools.change, () => this._build_tool_button_views())
+    this.connect(this.model.properties.tools.change, () => {
+      this._build_tool_button_views()
+      this.render()
+    })
     this.connect(this.model.properties.autohide.change, () => {
       this._toolbar_view_model.autohide = this.model.autohide
       this._on_visible_change()
@@ -143,7 +148,19 @@ export class ToolbarBaseView extends DOMView {
   }
 }
 
-export type GestureType = "pan" | "scroll" | "pinch" | "tap" | "doubletap" | "press" | "rotate" | "move" | "multi"
+export type GesturesMap = {
+  pan:       { tools: GestureTool[], active: Tool | null },
+  scroll:    { tools: GestureTool[], active: Tool | null },
+  pinch:     { tools: GestureTool[], active: Tool | null },
+  tap:       { tools: GestureTool[], active: Tool | null },
+  doubletap: { tools: GestureTool[], active: Tool | null },
+  press:     { tools: GestureTool[], active: Tool | null },
+  rotate:    { tools: GestureTool[], active: Tool | null },
+  move:      { tools: GestureTool[], active: Tool | null },
+  multi:     { tools: GestureTool[], active: Tool | null },
+}
+
+export type GestureType = keyof GesturesMap
 
 export namespace ToolbarBase {
   export type Attrs = p.AttrsOf<Props>
@@ -151,17 +168,7 @@ export namespace ToolbarBase {
   export type Props = Model.Props & {
     tools: p.Property<Tool[]>
     logo: p.Property<Logo>
-    gestures: p.Property<{
-      pan:       { tools: GestureTool[], active: Tool | null },
-      scroll:    { tools: GestureTool[], active: Tool | null },
-      pinch:     { tools: GestureTool[], active: Tool | null },
-      tap:       { tools: GestureTool[], active: Tool | null },
-      doubletap: { tools: GestureTool[], active: Tool | null },
-      press:     { tools: GestureTool[], active: Tool | null },
-      rotate:    { tools: GestureTool[], active: Tool | null },
-      move:      { tools: GestureTool[], active: Tool | null },
-      multi:     { tools: GestureTool[], active: Tool | null },
-    }>,
+    gestures: p.Property<GesturesMap>,
     actions: p.Property<ActionTool[]>
     inspectors: p.Property<InspectTool[]>
     help: p.Property<HelpTool[]>
@@ -171,6 +178,20 @@ export namespace ToolbarBase {
 }
 
 export interface ToolbarBase extends ToolbarBase.Attrs {}
+
+function createGestureMap(): GesturesMap {
+  return {
+    pan:       { tools: [], active: null },
+    scroll:    { tools: [], active: null },
+    pinch:     { tools: [], active: null },
+    tap:       { tools: [], active: null },
+    doubletap: { tools: [], active: null },
+    press:     { tools: [], active: null },
+    rotate:    { tools: [], active: null },
+    move:      { tools: [], active: null },
+    multi:     { tools: [], active: null },
+  }
+}
 
 export class ToolbarBase extends Model {
   properties: ToolbarBase.Props
@@ -189,17 +210,7 @@ export class ToolbarBase extends Model {
     })
 
     this.internal({
-      gestures: [ p.Any, () => ({
-        pan:       { tools: [], active: null },
-        scroll:    { tools: [], active: null },
-        pinch:     { tools: [], active: null },
-        tap:       { tools: [], active: null },
-        doubletap: { tools: [], active: null },
-        press:     { tools: [], active: null },
-        rotate:    { tools: [], active: null },
-        move:      { tools: [], active: null },
-        multi:     { tools: [], active: null },
-      })  ],
+      gestures:         [ p.Any,      createGestureMap ],
       actions:          [ p.Array,    []      ],
       inspectors:       [ p.Array,    []      ],
       help:             [ p.Array,    []      ],
@@ -208,6 +219,62 @@ export class ToolbarBase extends Model {
   }
 
   _proxied_tools?: (Tool | ToolProxy)[]
+
+  initialize(): void {
+    super.initialize()
+    this._init_tools()
+  }
+
+  protected _init_tools(): void {
+    // The only purpose of this function is to avoid unnecessary property churning.
+    const tools_changed = function (old_tools: Tool[], new_tools: Tool[]) {
+      if (old_tools.length != new_tools.length) {
+        return true
+      }
+      const new_ids = new Set(new_tools.map(t => t.id))
+      return some(old_tools, t=> !new_ids.has(t.id))
+    }
+    const new_inspectors = this.tools.filter(t => t instanceof InspectTool) as InspectTool[]
+    if (tools_changed(this.inspectors, new_inspectors)) {
+      this.inspectors = new_inspectors
+    }
+    const new_help = this.tools.filter(t => t instanceof HelpTool) as HelpTool[]
+    if (tools_changed(this.help, new_help)) {
+      this.help = new_help
+    }
+    const new_actions = this.tools.filter(t => t instanceof ActionTool) as ActionTool[]
+    if (tools_changed(this.actions, new_actions)) {
+      this.actions = new_actions
+    }
+    const check_event_type = (et: EventType, tool: Tool) => {
+      if (!(et in this.gestures)) {
+        logger.warn(`Toolbar: unknown event type '${et}' for tool: ${tool.type} (${tool.id})`)
+      }
+    }
+    const new_gestures = createGestureMap()
+    for (const tool of this.tools) {
+      if (tool instanceof GestureTool && tool.event_type) {
+        if (isString(tool.event_type)) {
+          new_gestures[tool.event_type].tools.push(tool)
+          check_event_type(tool.event_type, tool)
+        } else {
+          new_gestures.multi.tools.push(tool)
+          for (const et of tool.event_type) {
+            check_event_type(et, tool)
+          }
+        }
+      }
+    }
+    for (const et of Object.keys(new_gestures) as GestureType[]) {
+      const gm = this.gestures[et]
+      if (tools_changed(gm.tools, new_gestures[et].tools)) {
+        gm.tools = new_gestures[et].tools
+      }
+      if (gm.active && every(gm.tools, t => t.id != gm.active!.id)) {
+        gm.active = null
+      }
+    }
+  }
 
   get horizontal(): boolean {
     return this.toolbar_location === "above" || this.toolbar_location === "below"

--- a/bokehjs/src/lib/models/tools/toolbar_box.ts
+++ b/bokehjs/src/lib/models/tools/toolbar_box.ts
@@ -1,13 +1,10 @@
 import * as p from "core/properties"
 import {Location} from "core/enums"
-import {logger} from "core/logging"
-import {isString} from "core/util/types"
-import {some, includes, sort_by} from "core/util/array"
+import {includes, sort_by} from "core/util/array"
 
 import {Tool} from "./tool"
 import {ButtonTool} from "./button_tool"
 import {ActionTool} from "./actions/action_tool"
-import {HelpTool} from "./actions/help_tool"
 import {GestureTool} from "./gestures/gesture_tool"
 import {InspectTool} from "./inspectors/inspect_tool"
 import {ToolbarBase, GestureType} from "./toolbar_base"
@@ -35,46 +32,7 @@ export class ProxyToolbar extends ToolbarBase {
 
   initialize(): void {
     super.initialize()
-    this._init_tools()
     this._merge_tools()
-  }
-
-  protected _init_tools(): void {
-    for (const tool of this.tools) {
-      if (tool instanceof InspectTool) {
-        if (!some(this.inspectors, (t) => t.id == tool.id))
-          this.inspectors = this.inspectors.concat([tool])
-      } else if (tool instanceof HelpTool) {
-        if (!some(this.help, (t) => t.id == tool.id))
-          this.help = this.help.concat([tool])
-      } else if (tool instanceof ActionTool) {
-        if (!some(this.actions, (t) => t.id == tool.id))
-          this.actions = this.actions.concat([tool])
-      } else if (tool instanceof GestureTool) {
-        let event_types: GestureType[]
-        let multi: boolean
-        if (isString(tool.event_type)) {
-          event_types = [tool.event_type]
-          multi = false
-        } else {
-          event_types = tool.event_type || []
-          multi = true
-        }
-
-        for (let et of event_types) {
-          if (!(et in this.gestures)) {
-            logger.warn(`Toolbar: unknown event type '${et}' for tool: ${tool.type} (${tool.id})`)
-            continue
-          }
-
-          if (multi)
-            et = "multi"
-
-          if (!some(this.gestures[et].tools, (t) => t.id == tool.id))
-            this.gestures[et].tools = this.gestures[et].tools.concat([tool])
-        }
-      }
-    }
   }
 
   protected _merge_tools(): void {

